src/test/java/com/myroslav/cosmickitties/service/ProductControllerTest.java

```diff
@@ -0,0 +1,121 @@
+package com.myroslav.cosmickitties.service;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.myroslav.cosmickitties.dto.ProductDTO;
+import com.myroslav.cosmickitties.exception.GlobalExceptionHandler;
+import com.myroslav.cosmickitties.exception.ResourceNotFoundException;
+import com.myroslav.cosmickitties.service.ProductService;
+import com.myroslav.cosmickitties.controller.ProductController;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.math.BigDecimal;
+import java.util.List;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+/**
+ * Controller level tests using MockMvc and global exception handler.
+ * Tests positive and negative validation flows.
+ */
+class ProductControllerTest {
+
+    private MockMvc mockMvc;
+    private ProductService service;
+    private ObjectMapper objectMapper;
+
+    @BeforeEach
+    void setUp() {
```

Comment:
tha'ts difficult way to implement it, you are initializing the mock mvc - before each test, doing a lot of the "monkey job". You can configure the mock mvc with the @WebMvcTest, if you wanna do the unit test, you can do the @AutoconfigurableMockMvc to have the integration test
https://www.youtube.com/live/Nk39h-3Vvn4?si=6Y4RmgdNh7DOQqRG